domain/schema/model/sql/0007-charm.sql

```diff
@@ -1,27 +1,398 @@
+CREATE TABLE charm_run_as_kind (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_charm_run_as_kind_name
+    ON charm_run_as_kind (name);
+
+INSERT INTO charm_run_as_kind VALUES
+    (0, ''), 
+    (1, 'root'),
+    (2, 'sudoer'),
+    (3, 'non-root');
+
 CREATE TABLE charm (
-    uuid    TEXT PRIMARY KEY,
-    url     TEXT NOT NULL
+    uuid                TEXT PRIMARY KEY,
+    name                TEXT,
+    description         TEXT,
+    summary             TEXT,
+    min_juju_version    TEXT,
+    run_as_id           INT,
+    -- Assumes is a blob of YAML that will be parsed by the charm to compute
+    -- the result of the SAT expression.
+    -- As the expression tree is generic, you can't use RI or index into the
+    -- blob without constraining the expression to a specific set of rules.
+    assumes_blob        TEXT,
+    CONSTRAINT          fk_charm_run_as_kind_charm
+        FOREIGN KEY     (run_as_id)
+        REFERENCES      charm_run_as_kind(id)
+);
+
+CREATE UNIQUE INDEX idx_charm_name
+    ON charm (name);
+
+-- The charm_state table exists to store the availability of a charm. The
+-- fact that the charm is in the database indicates that it's a placeholder.
+-- Updating the available flag to true indicates that the charm is now
+-- available for deployment.
+-- This is exists as a separate table as the charm table models the charm
+-- metadata and the goal state of the charm. The charm_state table models the
+-- internal state of the charm.
+CREATE TABLE charm_state (
+    charm_uuid          TEXT NOT NULL,
+    -- Available is a flag that indicates whether the charm is available for
+    -- deployment.
+    available           BOOLEAN,
+    CONSTRAINT          fk_charm_state_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid)
+);
+
+CREATE TABLE charm_source (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_charm_source_name
+    ON charm_source (name);
+
+INSERT INTO charm_source VALUES
+    (0, 'local'),
+    (1, 'ch');
+
+CREATE TABLE charm_origin (
+    charm_uuid          TEXT NOT NULL,
+    source_id           INT,
+    id                  string,
+    hash                string,
```

Comment:
IIANM, we would use this hash to validate the downloaded blob. Do we need to store both this and the hash in the charm_hash table?

Reply:
Removed